packages/odyssey-react-mui/src/ThemeContrastProvider.tsx

```diff
@@ -0,0 +1,156 @@
+    const normalizedColor = normalizeRgbaToRgb(bgColor);
+    return normalizedColor === hueNeutral50Rgb
+      ? Tokens.HueNeutral50
+      : normalizedColor;
+  }
+  return bgColor === hueNeutral50Rgb ? Tokens.HueNeutral50 : bgColor;
+};
+
+/**
+ * Determines the effective background color of an element.
+ *
+ * @param element - The HTML element to check.
+ * @returns The effective background color. Returns "#ffffff" if no non-transparent background is found.
+ *
+ * Note:
+ * - Low contrast mode is used for white background (#ffffff or HueNeutralWhite).
+ * - High contrast mode is used for gray background (#f4f4f4 or HueNeutral50).
+ */
+export const getBackgroundColor = (element: HTMLElement | null): string => {
+  while (element) {
+    const bgColor = getElementComputedBackgroundColor(element);
+    if (!isTransparentColor(bgColor)) {
+      return normalizeBackgroundColor(bgColor);
+    }
+    element = element.parentElement;
+  }
+  return "#ffffff"; // Default to white/low contrast if no background color is found
+};
+
+type ThemeContrastProviderProps = {
+  children: ReactNode;
+  contrastMode?: ContrastMode;
+};
+
+export const ThemeContrastProvider: React.FC<ThemeContrastProviderProps> = ({
+  children,
+  contrastMode: explicitContrastMode,
+}) => {
+  const ref = useRef<HTMLDivElement>(null);
+  const [parentBackgroundColor, setParentBackgroundColor] = useState("#ffffff");
+  const [contrastMode, setContrastMode] = useState<ContrastMode>(
+    () => explicitContrastMode || "highContrast",
+  );
+
+  const updateBackgroundColor = useCallback(() => {
+    const newBgColor = getBackgroundColor(ref.current);
+    setParentBackgroundColor(newBgColor);
+
+    if (!explicitContrastMode) {
+      setContrastMode(
+        newBgColor === Tokens.HueNeutral50 ? "highContrast" : "lowContrast",
+      );
+    }
+  }, [explicitContrastMode]);
+
+  useLayoutEffect(() => {
+    const observer = new MutationObserver(updateBackgroundColor);
+    observer.observe(document.querySelector("html") as HTMLHtmlElement, {
+      attributes: true,
+      attributeFilter: ["class", "style"],
+    });
+    observer.observe(document.head, {
+      childList: true,
+      subtree: true,
+    });
+
+    const onTransitionEnd = (event: TransitionEvent) => {
+      if (event.propertyName === "background-color") {
+        updateBackgroundColor();
+      }
+    };
+
+    document.addEventListener("transitionend", onTransitionEnd);
+    updateBackgroundColor();
+
+    return () => {
+      document.removeEventListener("transitionend", onTransitionEnd);
+      observer.disconnect();
+    };
+  }, [updateBackgroundColor]);
+
+  const contextValue = useMemo(
+    () => ({
+      contrastMode,
+      parentBackgroundColor,
+    }),
+    [contrastMode, parentBackgroundColor],
+  );
+
+  return (
+    <div ref={ref}>
+      <ThemeContrastContext.Provider value={contextValue}>
+        {children}
+      </ThemeContrastContext.Provider>
+    </div>
```

Comment:
This should probably be `ThemeContrastModeContext` to match how we use it in other spots.